Annotated

phadej opened the pull request:
An attempt to solve #2.

There are some engineering questions:
- probably one should renamed `render` into `annotatedRender` and have `render` omit the annotations.
- another idea is to have `annotatedRender :: (a -> String -> r) -> d a -> r` where `r` is some "abstract" notion of concatenable strings (e.g. `Builder`, or `Html`); that will make `annotatedRender` much more useful.
- I boldly made the whole package annotated, but it might sense to provide some monomorphised functions (working on `Doc ()`), not sure which though.

*EDIT* I also noticed that `nest` (without prime) combinator is missing. Didn't read a paper to how to define it :) Btw, making use of `doctest` would make tracking correctness examples in the source manageable (there is an example in `annotate` haddocks).

*EDIT:* there is an obvious bug, as one should `mappend` annotations when mappending the `L` blocks.

jyp commented:
It looks like the "Ord" constraint on the annotation should go away: annotation should not influence any ordering. 

Also the changes to the merge algorithm in the commit 
https://github.com/jyp/prettiest/pull/4/commits/893c759bcc0ebf57003bb235f0c7eaf8c1c00e8b
seem to go in the wrong direction; they should go away as you remove this Ord constraint.

phadej commented:
Yes, I removed the `Ord` and `Semigroup` constraints.

phadej commented:
The rendered haddock example section

![image](https://cloud.githubusercontent.com/assets/51087/26584515/b54cb564-4551-11e7-852c-fac5857d1127.png)

*EDIT*, btw in current `paper` sources you have `abcd4` example output with **five** abcd blocks

jyp commented on Text/PrettyPrint/Compact/Core.hs:
```diff
@@ -1,83 +1,117 @@
+--
+-- First argument is length.
+--
+data AS a = AS !Int [(Maybe a, String)]
```
This data type needs to be explained and/or simplified:
- what are the invariants?
- why do we have a list? 
- why do you have a Maybe instead of using the mempty of a?

phadej replied:
I'll document the invariants

we need a list (I think, have to write a test), because `Semigroup L` instance concatenates the `AS` (previously `String`), and we have to preserve the individual piece annotations.

I'm not sure whether `List` or `Seq` or something else is better from performance stand point, but that's easy to change.

I think that absence of annotation is quite common, also makes possible to avoid cluttering ~everything with `Monoid a` constraint.

If we agree on these points, I'll add them as notes to the source.

phadej replied:
AS resembles the http://hackage.haskell.org/package/prettyprinter-1/docs/Data-Text-Prettyprint-Doc-Internal.html#t:SimpleDocStream quite a lot, except that `AS` cannot yet contain newlines.

jyp commented on Text/PrettyPrint/Compact/Core.hs:
```diff
@@ -1,83 +1,117 @@
+  annotatedRender :: Monoid r
+                  => (Maybe a -> String -> r) -- ^ how to annotate the string. /Note:/ the annotation should preserve the visible length of the string.
+                  -> d a                      -- ^ renderable
+                  -> r
```
why did you split annotatedRender in another class?

phadej replied:
Because `Render :: (* -> *) -> Constraint`, but `Layout :: * -> Constraint`.

jyp commented on Text/PrettyPrint/Compact/Core.hs:
```diff
@@ -88,46 +122,78 @@ pareto' acc (x:xs) = if any (≺ x) acc
 
+leFst :: Ord a => (a, b) -> (a, b) -> Bool
+leFst = (<=) `on` fst
 
```
What is the purpose of removing the Poset class? Please try to stick to one atomic change per PR.

phadej replied:
Good point.

If I understand the inner workings right, probably the better change would be

```diff
-bests :: forall a. (Ord a, Poset a) => [[a]] -> [a]
+bests :: forall a. (Ord a, Poset a) => [[(a, b)]] -> [(a, b)]
```
or 
```diff
-bests :: forall a. (Ord a, Poset a) => [[a]] -> [a]
+bestsOn :: forall a b. (Ord b, Poset b) => (a -> b) -> [[a]] -> [a]
```

This will let us remove questionable `Poset (a, b)` instance too, and we won't need `leFst`.

jyp commented:
I still believe that it is better to have the annotation be a monoid and drop the "Maybe".
Indeed: you assume that the annotations are composable (semigroup) and the "annotate" function makes extensive use of this fact (which is great!). In all cases that I can think of, these annotations will not only be a semigroup, but also a monoid (you give the example of uppercase flag; I can think of a set of flags, colors, etc.). Having "Maybe" here is thus redundant. One may think that "Maybe" saves some memory, but in reality it does not because the `mempty` value will be shared across all un-annotated texts.

Otherwise I think that I understand and agree with the rest of changes. Thanks!

phadej commented:
Changed `Maybe a` to `a`. Also added a benchmark suite in the package to check how refactorings affect performance.

```
benchmarking small/pretty-compact
time                 9.510 ms   (9.443 ms .. 9.582 ms)
                     1.000 R²   (0.999 R² .. 1.000 R²)
mean                 9.314 ms   (9.244 ms .. 9.363 ms)
std dev              158.6 μs   (119.3 μs .. 239.1 μs)

benchmarking small/pretty-compact Text
time                 10.30 ms   (10.13 ms .. 10.43 ms)
                     0.999 R²   (0.998 R² .. 1.000 R²)
mean                 10.22 ms   (10.16 ms .. 10.28 ms)
std dev              167.4 μs   (128.4 μs .. 238.8 μs)

benchmarking small/pretty
time                 1.907 ms   (1.893 ms .. 1.920 ms)
                     0.999 R²   (0.996 R² .. 1.000 R²)
mean                 1.911 ms   (1.890 ms .. 1.990 ms)
std dev              119.8 μs   (23.76 μs .. 268.5 μs)
variance introduced by outliers: 46% (moderately inflated)

benchmarking small/wl-pprint
time                 1.023 ms   (1.015 ms .. 1.029 ms)
                     0.999 R²   (0.998 R² .. 1.000 R²)
mean                 1.016 ms   (1.012 ms .. 1.021 ms)
std dev              13.71 μs   (10.88 μs .. 18.39 μs)
```

Seems that `Maybe` version is slightly faster, probably passing `Monoid` dictionary around slightly slows down building up the `Doc`.

> And makes our library about ten times as slow as that of Wadler-Leijen.

seems to be still true after these changes.

I experimented with using `DList` in `AS`, that got ~5% speedup; but I'll leave any further optimisations out of this PR.

phadej commented:
BTW, I noticed that the library is currently under GPL. If you later decide to relicense it as BSD3 or MIT, I'm totally ok with that too.

jyp commented:
Thanks! Merged.
